components/tooling/fetch-tests/src/main/java/mozilla/components/tooling/fetch/tests/FetchTestCases.kt

```diff
@@ -0,0 +1,416 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.tooling.fetch.tests
+
+import mozilla.components.concept.fetch.Client
+import mozilla.components.concept.fetch.MutableHeaders
+import mozilla.components.concept.fetch.Request
+import mozilla.components.concept.fetch.success
+import okhttp3.mockwebserver.MockResponse
+import okhttp3.mockwebserver.MockWebServer
+import okio.Buffer
+import okio.GzipSink
+import okio.Okio
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertTrue
+import org.junit.Assert.fail
+import org.junit.Test
+import java.io.File
+import java.io.IOException
+import java.net.ServerSocket
+import java.net.Socket
+import java.net.SocketTimeoutException
+import java.util.UUID
+import java.util.concurrent.TimeUnit
+
+/**
+ * Generic test cases for concept-fetch implementations.
+ *
+ * We expect any implementation of concept-fetch to pass all test cases here.
+ */
+@Suppress("IllegalIdentifier", "FunctionName", "unused")
+abstract class FetchTestCases {
+    abstract fun createClient(): Client
```

Comment:
🤔  Could this method introduce flakiness ? If the implementer always returns the same instance, this could be share between tests. Should we document this function about or change the method to createNewClient()?